### ISSUE TITLE
Tool to delete downloaded chapters that are not in the library

### ISSUE THREAD

**Issue by jasonhpchu** [inorichi/tachiyomi#1767]
I'm browsing the download folder for Tachiyomi, and noticed a number of old manga in there.
Some are probably orphaned from migration, or maybe I deleted a manga from the library without checking off the delete checkbox.

I suggest a clean up tool, perhaps a tool under Advanced, which will look for these and remove them.

Thank you.

**Comment by metroid112** [comment]
There's an option under Settings -> Advanced -> Clear database. That should do exactly what you want

**Comment by jasonhpchu** [comment]
Nope, tried, it, all those orphaned downloaded chapters remains in the download folder.
Try this, pick a manga, download a few chapters, migrate it to another source, download those chapter again.  Now you have two folders of the two source with those manga chapters downloaded twice.
Clear database doesn't clear one of those.

So overtime, you'll accumulate lots of these and wasting space.

**Comment by CarlosEsco** [comment]
Fyi clear database just clears database of manga not in your library.

**Comment by jasonhpchu** [comment]
^^That's what I thought, so my issue remains, a clean up tool would be nice.

Or perhaps migration should delete downloaded chapters from the source being migrated from.
But I'm sure there are other cases where a chapter might get orphaned.
Or maybe a reverse tool to scan the folder and add in entries for downloaded chapters that's not in the database, then the user can manually delete those entries which will also delete downloaded chapters.

**Comment by RED-404** [comment]
This is the workaround that I use.
Remove from library.
Click delete chapters at the bottom.
Click add to library.
[Example](https://youtu.be/nRLHs1-NVyE)

**Comment by jasonhpchu** [comment]
^^
Thanks for the comprehensive instructions.
However that only helps IF you already know if a few chapters of this manga from that source was orphaned.
For now, I'll always use Copy when doing a migration, then the Remove From Library with Delete downloaded will clean it up.

Excellent pointer about the Download badges, that'll show me orphaned stuff from manga still in the library, but hidden away because I filtered out read chapters.  Turning this on, I found one manga I had 90+ chapters downloaded!!  Thanks!

**Comment by Efreak** [comment]
A combination of a badge to indicate chapters that aren't in the database/cache, and the UI shown in #2079 would be useful

**Comment by LycorisNagami** [comment]
Clear Database is already a working function in app that clears manga history not in library so why not add into it code that checks if that manga has a folder in phone storage, if true then delete that folder. Perfect Housekeeping, no wasted phone storage

**Comment by KiloJuliett** [comment]
I think reasonable behaviour for a "clearing orphaned downloads" feature for the Clear Database button should be that for any valid manga from any valid source in the database *not* in the user's personal library, if that manga has downloaded chapters, that manga should be deleted from the filesystem. 

Some caveats that I think need further discussion: I **do not** believe it is a safe assumption that *any* orphaned download in Tachiyomi's downloads directory can be safety deleted (that is to say, it is *not necessarily the case* that any downloaded chapters that cannot be identified as a manga in the user's personal library should be deleted).

- It is possible that a source goes offline, in which case it is reasonable to expect users to want to save any downloaded chapters.
- It is possible that a source may "delete" a chapter, in which case once again users may want to keep an archive.
- Some user may be keeping personal files inside Tachiyomi's downloads directory. Perhaps they keep their personal collection of manga there, I dunno, but I can already see the words "Broken update deleted my files" appearing in the issues tab.
- How should automated clearing of orphaned downloads interact with the "Allow deleting bookmarked chapters" and "Excluded categories" options? Orphaned downloads are definitionally not in the user's library and so I think such chapters should be deleted, but this may be unexpected to other users.

Looking forward to hearing feedback.

**Comment by Efreak** [comment]
@KiloJuliett

> Some user may be keeping personal files inside Tachiyomi's downloads directory. Perhaps they keep their personal collection of manga there, I dunno, but I can already see the words "Broken update deleted my files" appearing in the issues tab.

The solution here to tell users not to store files in the Tachiyomi downloads folder. That's what the local folder is for.

> How should automated clearing of orphaned downloads interact with the "Allow deleting bookmarked chapters" and "Excluded categories" options? Orphaned downloads are definitionally not in the user's library and so I think such chapters should be deleted, but this may be unexpected to other users.

Isn't this already dealt with? Tachiyomi SY provides checkboxes for orphaned, read, and not in library. I don't think excluded chapters should be dealt with here--if you really want excluded chapters to be saved, you should delete things manually; I recommend using an app like DiskUsage or a file manager to identify/remove directories (titles) taking up large amounts of space

**Comment by cecoates** [comment]
I recently had to reinstall my OS and restore a Tachiyomi backup. Pointing my Downloads folder to the one that I had backed up from my previous install, you do end up with mismatches. If you don't remember everything you previously downloaded, or check each source/folder individually, you can easily download duplicates.

Even if you're not reinstalling you can end up with a bit of clutter over time. Chapters might change names at the source, for example, so you can end up redownloading items you technically already have.

A tool to clean up the directory would be pretty handy.

**Comment by recheej** [comment]
Hi there. I'm currently taking a look at this issue. My current plan is to grab all the non-library mangas from the library and then allow you to just delete all of the folders in the download directory related to those mangas. Let me know if there are any questions/comments. This is my first issue in this repository.

https://github.com/recheej/tachiyomi/tree/rejozil/deleteorphanedmangas

**Comment by arkon** [comment]
You can have leftover downloads from things that were cleared from the database though. You'd probably want to list the folder names and check against what's actually in the library.

**Comment by recheej** [comment]
> You can have leftover downloads from things that were cleared from the database though. You'd probably want to list the folder names and check against what's actually in the library.

Ah I see. Should clear database be the same thing as what's being requested in this issue? I'm new to the app but from looking at the comments from various issues, seems like there's some confusion on what clear database does (some think it will delete the files too).

**Comment by arkon** [comment]
No, I'd keep the functionality separate, at least for now. We can revisit it if there's feedback otherwise.

**Comment by recheej** [comment]
> No, I'd keep the functionality separate, at least for now. We can revisit it if there's feedback otherwise.

Got it. Just so we're on the same page, could you explain the difference (from the user's perspective) between clear database and clear orphaned downloads? In clear DB we also remove history I see.

Also, should I add code to delete old source chapters upon migration?

**Comment by arkon** [comment]
> could you explain the difference (from the user's perspective) between clear database and clear orphaned downloads? In clear DB we also remove history I see.

I'm not entirely sure if I can speak for how a typical user interprets it since I'd have a "proper" understanding of what they're doing. Clearing database clears manga entries, which implicitly removes their corresponding history entries. Clearing orphaned downloads is relevant for things that can be finer grain than removed manga entries, like removed chapters.

> Also, should I add code to delete old source chapters upon migration?

That was added in #9621

**Comment by recheej** [comment]
> > could you explain the difference (from the user's perspective) between clear database and clear orphaned downloads? In clear DB we also remove history I see.
> 
> I'm not entirely sure if I can speak for how a typical user interprets it since I'd have a "proper" understanding of what they're doing. Clearing database clears manga entries, which implicitly removes their corresponding history entries. Clearing orphaned downloads is relevant for things that can be finer grain than removed manga entries, like removed chapters.
> 
> > Also, should I add code to delete old source chapters upon migration?
> 
> That was added in #9621

I see! Thanks for that info! Another question I have is what are the situations in the latest version of the app where downloads are orphaned (app downloaded file, downloaded file isn't needed anymore, and user doesn't know)? I'd ask questions on the discord but my account currently isn't working :(

**Comment by arkon** [comment]
Typically if the source changes its list of chapters so previously existing and downloaded chapter(s) get orphaned.
